reconcile/utils/secret_reader.py

```diff
@@ -27,8 +27,107 @@ class SupportsSecret(Protocol):
     q_format: Optional[str]
 
 
+class SupportsVaultSettings(Protocol):
+    """
+    SupportsVaultSettings defines all attributes needed from app-interface-settings
+    to instantiate a SecretReader
+    """
+
+    vault: bool
+
+
+class TypedSecretReader:
+    def read(self, secret: SupportsSecret) -> dict[str, str]:
+        raise NotImplementedError()
+
+    def read_all(self, secret: SupportsSecret) -> dict[str, str]:
+        raise NotImplementedError()
+
+    def _secret_to_dict(self, secret: SupportsSecret) -> dict[str, Any]:
+        """
+        Config.read() and VaultClient.read() do not support types yet.
+        Once they do, we can remove this helper function.
+        """
+        return {
+            "path": secret.path,
+            "field": secret.field,
+            "version": secret.version,
+            "format": secret.q_format,
+        }
+
+
+class VaultSecretReader(TypedSecretReader):
+    """
+    Read secrets from vault via a vault_client
+    """
+
+    def __init__(self, vault_client: Optional[VaultClient] = None):
+        self._vault_client = vault_client
+
+    @property
+    def vault_client(self):
+        if self._vault_client is None:
+            self._vault_client = VaultClient()
+        return self._vault_client
+
+    @retry()
+    def read(self, secret: SupportsSecret) -> dict[str, str]:
+        try:
+            data = self.vault_client.read(self._secret_to_dict(secret))
+        except vault.SecretNotFound as e:
+            raise SecretNotFound(*e.args) from e
+        return data
+
+    @retry()
+    def read_all(self, secret: SupportsSecret) -> dict[str, str]:
+        try:
+            data = self.vault_client.read_all(self._secret_to_dict(secret))
+        except Forbidden:
+            raise VaultForbidden(
+                f"permission denied reading vault secret " f"at {secret.path}"
+            )
+        except vault.SecretNotFound as e:
+            raise SecretNotFound(*e.args) from e
+        return data
+
+
+class ConfigSecretReader(TypedSecretReader):
+    """
+    Read secrets from a config file
+    """
+
+    def read(self, secret: SupportsSecret) -> dict[str, str]:
+        try:
+            data = config.read(self._secret_to_dict(secret))
+        except config.SecretNotFound as e:
+            raise SecretNotFound(*e.args) from e
+        return data
+
+    def read_all(self, secret: SupportsSecret) -> dict[str, str]:
+        try:
+            data = config.read_all(self._secret_to_dict(secret))
+        except config.SecretNotFound as e:
+            raise SecretNotFound(*e.args) from e
+        return data
+
+
+def create_secret_reader(
+    settings: Optional[SupportsVaultSettings],
```

Comment:
Thinking about this even more, could this even be `def create_secret_reader(vault: Optional[bool] = None)`. As per the points above, this should support settings whether it's a `dict`, `SupportsVaultSettings`, or if someone has some other use case? It would only require that you make a minor change to the calls:

```python
create_secret_reader(vault=settings.vault)
# Instead of
create_secret_reader(settings)
```

Reply:
Adjusted now as you suggest :+1: